app/domain/authentication/authn_azure/validate_application_identity.rb

```diff
@@ -0,0 +1,102 @@
+      # TODO For a Conjur user, an incorrect full id will be returned
+      def role_id
+        @role_id ||= @role_class.roleid_from_username(@account, @username)
+      end
+
+      def role
+        @role ||= @resource_class[role_id]
+        raise SecurityErr::RoleNotFound(role_id) if @username.nil?
+        return @role
+      end
+
+      # TODO add parsing format to check xms mirid is in proper format
+      # xms_mirid is a term in Azure to define a claim that describes the resource that holds the encoding of the instance's
+      # subscription, resource group, and provider identity needed for authorization. xms_mirid is one of the fields in
+      # the JWT token. This function will extract the relevant information from xms_mirid claim and populate a representative
+      # hash with the appropriate fields.
+      def parse_xms_mirid
+        # xms_mirid claim starts with an extra '/' so we will be ignoring it for parsing purposes
+        # ex: /subscription/<subscription_id> -> subscription/<subscription_id>
+        _, *field_split = @xms_mirid.split('/')
+        @xms_mirid_hash = Hash[field_split.each_slice(2).to_a]
+
+        @token_identity = {
+            subscription_id: @xms_mirid_hash["subscriptions"],
+            resource_group: @xms_mirid_hash["resourcegroups"]
+        }
+
+        # determines which Azure assigned identity is provided in annotations
+        # user-assigned-identity:
+        #   - validates that the correct provider (Microsoft.ManagedIdentity) for a user identity is defined in the
+        #     claim. If so, the 'user_assigned_identity' attribute will be added to the hash with the corresponding
+        #     value from xms_mirid claim
+        # system-assigned-identity:
+        #   - validates that the correct provider (Microsoft.Compute) for a system identity is defined in the
+        #     claim and its resource is one that we support. At current, we only support a virtualMachines resource.
+        #     If these conditions are met, a 'system_assigned_identity' attribute will be added to the hash with its
+        #     value being the oid field in the JWT token.
+        if @xms_mirid_hash["providers"] == "Microsoft.ManagedIdentity"
+          @logger.debug(Log::AuthnAzure::ExtractingIdentityForAuthentication.new(@xms_mirid_hash["providers"] << "/" << @xms_mirid_hash.keys[-1]))
+          @token_identity[:user_assigned_identity] = @xms_mirid_hash["userAssignedIdentities"]
+        else
+          @logger.debug(Log::AuthnAzure::ExtractingIdentityForAuthentication.new(@xms_mirid_hash["providers"] << "/" << @xms_mirid_hash.keys[-1]))
+          @token_identity[:system_assigned_identity] = @oid
+        end
+      end
+
+      # validate the integrity of host annotations against the xms_mirid object representation
+      def validate_application_identity
+        @logger.debug(Log::AuthnAzure::ValidatingApplicationIdentity.new(@xms_mirid_hash[@xms_mirid_hash.keys[-1]]))
```

Comment:
seems pretty weird to still use `@xms_mirid_hash` after you transformed it to `token_identity`. Please see how you can use `token_identity` here as well. It's not clear what `@xms_mirid_hash[@xms_mirid_hash.keys[-1]]` represents.

Reply:
A couple of ideas as to why I am against this:
1. `xms_mirid_hash` is a direct copy of what is decoded and verified and `token_identity` object is my own object representation with manipulations on the `xms_mirid` claim. Therefore, I am wary to say `xms_mirid_hash` == `token_identity`... UNLESS we address the following:
2. The `token_identity` object doesn't have an entry for the resource name when Azure resource has a system-assigned-identity. We can definitely add it to the `token_identity` w.o harming the logic because we are comparing what are in the annotations against what is the `token_identity` and not the inverse
3. `@xms_mirid_hash[@xms_mirid_hash.keys[-1]]` I agree that this is a bit ugly and my goal is for this to be the last entry in the xms_mirid claim (the resource name). For example: `sigalVM`

Reply:
You log that you are validating `sigalVM` but you are actually validating the `oid`.

Reply:
I prefer adding in to the token identity rather than to continue referring `xms_mirid_hash`. `xms_mirid_hash` should be a temporary object that helps you move from the `xms_mirid` string to a token identity.

Reply:
wasn't sure if oid was sensitive info. So you are suggesting that we have an extra attribute called `resource_name` where for user-identity it is the resource-name (the last parm in the xms_mirid claim) and for system-identity, oid is returned?

Reply:
My thought process was the following:
- The last index, `sigalVMUbuntu` in the xms_mirid claim would be descriptive enough for the admin to understand which VM is being referred to without exposing too much info about the VM. The same for the user-assigned id

Reply:
let's not get into what is sensitive and what is not for now. I asked @shaharglazner about the `xms_mirid` and `oid`. anyway, you can't tell the user that your are validating `sigalVM` while you are actually validating the `oid`. any part of the `xms_mirid` that is useful for us you can put in the `TokenIdentity`